### ISSUE TITLE
rename CQ config options

### ISSUE THREAD

**Issue by beckettsean** [influxdb/influxdb#5136]
Instead of 

```
compute-runs-per-interval = 10
compute-no-more-than = "2m0s"
recompute-previous-n = 2
recompute-no-older-than = "10m0s"
```

I propose

```
incremental-queries-per-interval = 10
min-time-between-batches = "2m0s"
max-resampled-queries-per-batch = 2
max-age-of-queried-data = "10m0s"
```

@rkuchan @gunnaraasen @dgnorton thoughts, improvements, or suggestions?

**Comment by rkuchan** [comment]
🎉👍🎉

**Comment by dgnorton** [comment]
I don't think changing the names makes them easier to understand but don't have strong feelings about it if everyone else agrees that the change makes it easier for users.  Also, @pauldix my veto breaking config change.

I think the real solution is to rethink how it computes when CQs should run.  It might be simpler for users if the query language allowed per-CQ settings.  I think we could eliminate two of the four settings that way.  E.g.,
```
CREATE CONTINUOUS QUERY mycq ON mydb
   PREVIOUS 1
   PERINTERVAL 1
   BEGIN
      SELECT mean(value) INTO cpu_mean FROM cpu GROUP BY time(1m)
   END
```
The parser could default `PREVIOUS` to `1` and `PERINTERVAL` to `1`.  For the example above, the query would run once for the current and previous interval.  That makes the query simpler to reason about and defaulting to `1` for both of those would reduce the chances of an unexpectedly expensive CQ.

Another example:
```
CREATE CONTINUOUS QUERY mycq ON mydb
   PERINTERVAL 4
   BEGIN
      SELECT mean(value) INTO cpu_mean FROM cpu GROUP BY time(1h)
   END
```
In this example, we've taken the default `PREVIOUS 1` and just specified how many times we want this CQ to run for the current interval and previous interval.  `1h / 4 = 15m`...so it will run once every 15 minutes.

Even better would be running only the CQs whose intervals have changed.  Then we probably wouldn't need any settings.  But that's a hard problem.

**Comment by pauldix** [comment]
I agree with @dgnorton's suggestions. All of these settings should be pulled from the config file in favor of a couple of easier to understand settings tied to each CQ. This goes inline with our use of config files: they're for host settings only. Anything that is applicable in the cluster should be in the metastore (thus the config should be specified via the query language).

I think that @dgnorton's two suggested settings cover all the bases and it's much more flexible since it's on a per CQ basis. What do you guys think?

**Comment by gunnaraasen** [comment]
I'm strongly in favor of @dgnorton's suggestion to set these config options per CQ.

Would it make sense to decouple the runs per interval setting for the current and previous intervals? Say there's the following CQ.

```
CREATE CONTINUOUS QUERY mycq ON mydb
   PREVIOUS 3
   PERINTERVAL 60
   BEGIN
      SELECT mean(value) INTO cpu_mean FROM cpu GROUP BY time(1h)
   END
```

The goal would be to have the latest CQ number up to date every minute. However, this would also end up recomputing the previous period every minute when we maybe only want it to recompute once or four times. Not sure if that's a real issue people will run into or even an issue that's worth solving.

Also, if we're going to move per-CQ scheduling out of the config, I think it makes sense to make the continuous querier service `RunInterval` configurable. `RunInterval` is currently hard-coded to one second, which is unnecessarily high for some low power/perf machines.

**Comment by jsternberg** [comment]
I have laid out the initial ground work for allowing the scanner and parser to understand the currently proposed syntax and pushed it to the `js-5136-per-cq-options` branch.

I did run into one small issue with the use of `PREVIOUS`. The ident string "previous" is already used in one place for the `fill(previous)` function. When `PREVIOUS` gets turned into a reserved keyword, this breaks `fill(previous)`. I was able to work around the issue by teaching the `ParseExpr` function to treat the `PREVIOUS` token in an expression context differently than in other contexts, but that may be too hacky of a solution. Is there a better name we can use?

**Comment by janantala** [comment]
It would be great to have these settings per CQ, like @dgnorton 's suggestion :+1:

**Comment by jsternberg** [comment]
I've been thinking about what the name of the token should be. I don't think the word `PREVIOUS` helps describe what the option does. To my understanding, the option is supposed to change the number of previous intervals to resample. Maybe we should consider using `RESAMPLE` instead? The word `PREVIOUS` can refer to anything related to previous results, while `RESAMPLE` seems to be more specific to me.

@dgnorton @pauldix any thoughts on what word we should use in the syntax?

**Comment by pauldix** [comment]
`RESAMPLE` works for me

**Comment by dgnorton** [comment]
`RESAMPLE` isn't clear to me.  It doesn't necessarily mean we're going into the past and (to me) almost sounds like a better replacement for `PERINTERVAL`.  How about...

`INTERVALS`
`PREVIOUS INTERVALS`
`RESAMPLE INTERVALS`
`PAST INTERVALS`
`PAST`
`PAST GROUPS`
`GROUPS`
?

**Comment by pauldix** [comment]
Yeah, now that you mention it, `RESAMPLE` works better for the current interval.

Having a two word phrase feels clunky to me too. Although I don't have any better suggestions.

**Comment by gunnaraasen** [comment]
How about changing the options from integers to durations? Then the syntax could be something like `RESAMPLE EVERY 10m FOR 2h`.

**Comment by jsternberg** [comment]
I think `RESAMPLE EVERY 10m FOR 2h` might be the best option. It seems to describe exactly what this does and is easier to read than the other options.

It might require too many reserved keywords though. @pauldix @dgnorton thoughts?

**Comment by pauldix** [comment]
@gunnaraasen, @jsternberg I like it. As long as the `EVERY` or `FOR` clauses are optional. Well, at least one must be present if you have the `RESAMPLE`.

**Comment by dgnorton** [comment]
I like `RESAMPLE EVERY 10m FOR 2h` but how should it handle it if the `FOR` duration isn't a multiple of the `GROUP BY` duration?

**Comment by jsternberg** [comment]
@dgnorton I was just going to not check for that and the math would just end up working itself out. So if you had `RESAMPLE EVERY 2m FOR 9m`, it would resample on `2m`, `4m`, `6m`, and `8m`, but not `10m`.

Likely worth a test case though to ensure that's true.

**Comment by gunnaraasen** [comment]
@dgnorton @jsternberg I vote for being inclusive to the nearest interval. That way you know at least all CQ-generated measurements are correct for data written within the `FOR` duration. Applied to @jsternberg's example, `RESAMPLE EVERY 2m FOR 9m` would resample the `0m-2m`, `2m-4m`, `4m-6m`, `6m-8m`, __and__ `8m-10m` intervals.

**Comment by jsternberg** [comment]
I've pushed the latest on this to `js-5136-per-cq-options`. It includes what I believe to be a working implementation. It doesn't do error checking to see if the `EVERY` and `FOR` values make any sense, but it should still do the correct thing anyway.

I don't think we should have to deal with strange durations as we'll provide some error messages around them instead to prevent the kinds of edge cases we're talking about here, but for now, `RESAMPLE EVERY 2m FOR 9m` would resample on the 2, 4, 6, and 8 minute intervals.

**Comment by pauldix** [comment]
can you create a PR so we can comment on it there?

**Comment by jsternberg** [comment]
Pushed as part of #5194. It still needs unit tests and then likely a cleanup of the stuff that isn't relevant anymore, at least for a first iteration.

**Comment by jsternberg** [comment]
Merged. This will be included in 0.10.0.
